src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4927,46 +4927,61 @@ bool StorageReplicatedMergeTree::getFakePartCoveringAllPartsInPartition(const St
 
     zkutil::ZooKeeperPtr zookeeper = getZooKeeper();
+    LogEntry entry;
 
+    dropPartImpl(zookeeper, part_name, entry, false, false);
+}
+
+void StorageReplicatedMergeTree::dropPart(const String & part_name, bool detach, ContextPtr query_context)
+{
+    assertNotReadonly();
+    if (!is_leader)
+        throw Exception("DROP PART cannot be done on this replica because it is not a leader", ErrorCodes::NOT_A_LEADER);
+
+    zkutil::ZooKeeperPtr zookeeper = getZooKeeper();
     LogEntry entry;
-    bool did_drop;
 
-    if (drop_part)
-    {
-        String part_name = partition->as<ASTLiteral &>().value.safeGet<String>();
-        did_drop = dropPart(zookeeper, part_name, entry, detach, throw_if_noop);
-    }
-    else
-    {
-        String partition_id = getPartitionIDFromQuery(partition, query_context);
-        did_drop = dropAllPartsInPartition(*zookeeper, partition_id, entry, query_context, detach);
-    }
+    bool did_drop = dropPartImpl(zookeeper, part_name, entry, detach, true);
 
     if (did_drop)
     {
         /// If necessary, wait until the operation is performed on itself or on all replicas.
-        if (query_context->getSettingsRef().replication_alter_partitions_sync != 0)
-        {
-            if (query_context->getSettingsRef().replication_alter_partitions_sync == 1)
-                waitForReplicaToProcessLogEntry(replica_name, entry);
-            else
-                waitForAllReplicasToProcessLogEntry(entry);
-        }
+        if (query_context->getSettingsRef().replication_alter_partitions_sync == 1)
+            waitForReplicaToProcessLogEntry(replica_name, entry);
+        else if (query_context->getSettingsRef().replication_alter_partitions_sync == 2)
+            waitForAllReplicasToProcessLogEntry(entry);
     }
+}
 
-    if (!drop_part)
+void StorageReplicatedMergeTree::dropPartition(const ASTPtr & partition, bool detach, ContextPtr query_context)
+{
+    assertNotReadonly();
+    if (!is_leader)
+        throw Exception("DROP PARTITION cannot be done on this replica because it is not a leader", ErrorCodes::NOT_A_LEADER);
+
+    zkutil::ZooKeeperPtr zookeeper = getZooKeeper();
+    LogEntry entry;
+
+    String partition_id = getPartitionIDFromQuery(partition, query_context);
+    bool did_drop = dropAllPartsInPartition(*zookeeper, partition_id, entry, query_context, detach);
+
+    if (did_drop)
     {
-        String partition_id = getPartitionIDFromQuery(partition, query_context);
-        cleanLastPartNode(partition_id);
+        /// If necessary, wait until the operation is performed on itself or on all replicas.
+        if (query_context->getSettingsRef().replication_alter_partitions_sync == 1)
+            waitForReplicaToProcessLogEntry(replica_name, entry);
+        else if (query_context->getSettingsRef().replication_alter_partitions_sync == 2)
+            waitForAllReplicasToProcessLogEntry(entry);
     }
+
+    cleanLastPartNode(partition_id);
```

Comment:
looked bit safer inside `if (did_drop)` (why should we clean it if the drop was not successful?)